API doc revisions

davidtheclark opened the pull request:
Ref #2749

@lucaswoj I was thinking I could gradually build up the revisions in this PR, working a little bit each day, and ask questions as I go. Sound good?

davidtheclark commented on js/ui/map.js:
```diff
@@ -57,47 +57,51 @@ var defaultOptions = {
- * @param {number} [options.bearing] The bearing (rotation) of the map's initial viewport measured in degrees counter-clockwise from north.
- * @param {number} [options.pitch] The pitch of the map's initial viewport measured in degrees.
+ * @param {boolean} [options.preserveDrawingBuffer=false] If `true`, the map's canvas can be exported to a PNG using `map.getCanvas().toDataURL();`. This is `false` by default as a performance optimization.
+ * @param {LngLatBounds|Array<Array<number>>} [options.maxBounds] If set, the map is constrained to the given bounds. // TODO: Can LngLatBounds be like LngLat, standing in for Array<Array<string>>
```
My understanding is that throughout the docs `@param {LngLat}` stands in for `@param {LngLat | Array<string>}`. Does it make sense to do the same for `LngLatBounds` and `Array<Array<number>>`?

lucaswoj replied:
Yes! That makes sense!

Another option is to create a `LngLatLike` `@typedef`. This would allow us to keep the docs more concise and add new supported formats in the future. 

```
 * @typedef {(LngLat | Array<string>)} LngLatLike
```

davidtheclark replied:
That is a great idea. I'll try to figure out how to make that work with the generated page.

davidtheclark commented on js/ui/map.js:
```diff
@@ -57,47 +57,51 @@ var defaultOptions = {
+ * @param {LngLat} [options.center] The geographic coordinate on which the map's initial viewport is centered. TODO: What is default?
+ * @param {number} [options.zoom] The zoom level of the map's initial viewport. TODO: What is default?
+ * @param {number} [options.bearing] The bearing (rotation) of the map's initial viewport, measured in degrees counter-clockwise from north. TODO: What is default?
+ * @param {number} [options.pitch] The pitch of the map's initial viewport, measured in degrees. TODO: What is default?
```
What are the definitive default values for center, zoom, bearing and pitch? Is it accurate to just say 0 for all of them?

lucaswoj replied:
- center: [0, 0]
 - zoom: 0
 - bearing: 0
 - pitch: 0

jfirebaugh replied:
We should note though that the style can also specify these values. In descending order of precedence:

* Explicitly-specified option to `Map` constructor
* Specified in style
* "0" defaults

lucaswoj commented on js/ui/map.js:
```diff
@@ -57,47 +57,51 @@ var defaultOptions = {
- * @param {number} [options.maxZoom=20] Maximum zoom of the map
- * @param {Object|string} [options.style] Map style. This must be an an object conforming to the schema described in
- * the [style reference](https://mapbox.com/mapbox-gl-style-spec/), or a URL to a JSON style. To load a style from the
+ * @param {Element|string} options.container The HTML element to initialize the map in, or the element's string `id`.
```
>  - `@param`, `@property`, and `@returns` descriptions should be capitalized and end with a period. They should begin as if completing a sentence beginning with "This is..." or "This..."

What do you think about revising the standard to make these start with a lowercase letter? They aren't complete sentences so formatting them as such is odd.

davidtheclark replied:
I understand the reaction, but I think it's actually a pretty standard format for definitions and reference material. Many dictionary or encyclopedia, I think, will start most entries with fragments that begin with capital letters and end with periods. We could go the other way — always lowercase — but then we end up being inconsistent when a sentence happens to be complete, or we have a fragment followed by full sentences. What do you think?

davidtheclark replied:
BTW totally fine if you want to go to lowercase, was just explaining the rationale.

lucaswoj replied:
@tmcw, do you have any thoughts?

tmcw replied:
Yeah, I think I agree with @davidtheclark's viewpoint here. Though they are sentence fragments, having them start with caps and end with periods works functionally in this setting.

lucaswoj replied:
Great! I'm on board with this decision 🚀.

lucaswoj commented:
This is so so so great! Thank you for spearheading this project @davidtheclark.

davidtheclark commented on js/ui/map.js:
```diff
@@ -57,47 +57,61 @@ var defaultOptions = {
+
+/**
+ * A [`LngLatBounds`](#LngLatBounds) object or an array of [`LngLatLike`](#LngLatLike) objects.
+ * @typedef {(LngLatBounds | Array<LngLatLike>)} LngLatBoundsLike
```
Added these `@typedef`s, but not sure how to get them to render in a good place via documention.js. Seems like maybe they aren't being parsed correctly, either ... or my syntax is off. @tmcw Any insight here?

lucaswoj commented:
All the latest commits look great! 🏆

davidtheclark commented:
Cool. I will continue in that pattern, then.

davidtheclark commented:
I finished going through all the documentation, at last, and rebased the work into a single commit. I do have a few outstanding questions:

- Am I right in thinking that pitch is on a scale of 0-60?
- I become uncertain about which events provide EventData objects and which just provide the original DOM event? And are there EventData objects that do not contain an originalEvent, or is EventData only ever produced as the result of an interaction?

davidtheclark commented:
Anybody in @mapbox/copycop interested in helping fix up the Mapbox GL JS docs? All feedback welcome on the changes in this PR!

lucaswoj commented on docs/_plugins/insert_token.rb:
```diff
@@ -1,6 +1,6 @@
 module InsertToken
     def insert_token(input)
-        input.sub("<script>", "<script>\nmapboxgl.accessToken = '<your access token here>';")
+        input.sub("<script>", "<script>\n  mapboxgl.accessToken = '<your access token here>';")
```
Why did you add 2 spaces here?

The formatting of our examples suggests this line should not be indented. For example, see https://www.mapbox.com/mapbox-gl-js/examples

davidtheclark replied:
Ah ... the docs page is different (screenshot below). But instead of changing what I changed, we should probably just unindent the `map` line there so it matches the examples, right?

![screen shot 2016-06-28 at 2 17 49 pm](https://cloud.githubusercontent.com/assets/628431/16432696/2bdb406c-3d3b-11e6-9a39-1dd5c75d1672.png)

lucaswoj replied:
> we should probably just unindent the map line there so it matches the examples, right?

Sounds good 👍

lucaswoj commented on js/geo/lng_lat_bounds.js:
```diff
@@ -5,12 +5,18 @@ module.exports = LngLatBounds;
+ *
+ * Note that any Mapbox GL method that accepts a `LngLatBounds` object as an argument or option
+ * can also accept an `Array` of two [`LngLatLike`](#LngLatLike) constructs and will perform an implicit conversion.
+ * This flexible type is documented as [`LngLatLikeBounds`](#LngLatLikeBounds).
```
`LngLatLikeBounds` --> `LngLatBoundsLike`

lucaswoj commented on js/source/geojson_source.js:
```diff
@@ -86,9 +89,9 @@ GeoJSONSource.prototype = util.inherit(Evented, /** @lends GeoJSONSource.prototy
 
     /**
-     * Update source geojson data and rerender map
+     * Sets the GeoJSON data and re-render the map.
```
`re-render` -> `re-renders`

lucaswoj commented on js/source/geojson_source.js:
```diff
@@ -10,16 +10,19 @@ var EXTENT = require('../data/bucket').EXTENT;
+ * @param {Object|string} options.data A GeoJSON data object or a URL to one. The latter is preferable in the case of large GeoJSON objects.
+ * @param {number} [options.maxzoom=18] The maximum zoom level at which to preserve detail (1-20).
+ * @param {number} [options.buffer] The tile buffer, measured in pixels.
+ * @param {number} [options.tolerance] The simplification tolerance, measured in pixels (higher means simpler).
```
Would it be possible to link to a more thorough explanation of this parameter? For example, the wikipedia page https://en.wikipedia.org/wiki/Ramer%E2%80%93Douglas%E2%80%93Peucker_algorithm

lucaswoj commented on js/source/geojson_source.js:
```diff
@@ -10,16 +10,19 @@ var EXTENT = require('../data/bucket').EXTENT;
- * @param {number} [options.clusterMaxZoom] Max zoom to cluster points on. Defaults to one zoom less than `maxzoom` (so that last zoom features are not clustered).
+ * @param {Object|string} options.data A GeoJSON data object or a URL to one. The latter is preferable in the case of large GeoJSON objects.
+ * @param {number} [options.maxzoom=18] The maximum zoom level at which to preserve detail (1-20).
+ * @param {number} [options.buffer] The tile buffer, measured in pixels.
```
Would it be possible to link to a more thorough explanation of this parameter? This documentation is the best I could find https://www.mapbox.com/vector-tiles/specification/#clipping

cc @jakepruitt @mapsam

mapsam replied:
Hmmm, I wish http://mapnik.org/documentation/node-mapnik/3.5/#VectorTile described it better, but it doesn't. I have some things to update to the mapnik docs, and can include this.

lucaswoj commented on js/source/image_source.js:
```diff
@@ -15,11 +15,15 @@ var VertexArrayObject = require('../render/vertex_array_object');
- * @param {string} options.url A string URL of an image file
- * @param {Array} options.coordinates Four geographical [lng, lat] coordinates in clockwise order defining the corners (starting with top left) of the image. Does not have to be a rectangle.
+ * @param {string} options.url The URL of an image file.
+ * @param {Array} options.coordinates Four geographical coordinates,
```
Would it be accurate to list the type signature as `Array<Array<number>>`?

lucaswoj commented on js/source/image_source.js:
```diff
@@ -64,9 +68,12 @@ ImageSource.prototype = util.inherit(Evented, /** @lends ImageSource.prototype *
+     * Sets the image's coordinates and re-renders the map.
      *
-     * @param {Array} coordinates Four geographical [lng, lat] coordinates in clockwise order defining the corners (starting with top left) of the image. Does not have to be a rectangle.
+     * @param {Array} coordinates Four geographical coordinates,
```
Would it be accurate to list the type signature as `Array<Array<number>>`?

lucaswoj commented on js/source/video_source.js:
```diff
@@ -73,9 +77,9 @@ VideoSource.prototype = util.inherit(Evented, /** @lends VideoSource.prototype *
+     * Returns the HTML `video` element.
      *
-     * @returns {Object}
+     * @returns {Node} The HTML `video` element.
```
I think documentation.js plans to leverage Flow's type checker. The type Flow would use here is `HTMLVideoElement`.

lucaswoj commented on js/ui/handler/scroll_zoom.js:
```diff
@@ -146,31 +151,31 @@ ScrollZoomHandler.prototype = {
- * Zoom start event. This event is emitted just before the map begins a transition from one
- * zoom level to another, either as a result of user interaction or the use of methods such as `Map#jumpTo`.
+ * Fired just before the map begins a transition from one zoom level to another,
+ * as the result of either user interaction or methods such as [`Map#jumpTo`](#Map#jumpTo).
```
💭 Because `Map#jumpTo` is instantaneous, I don't think it triggers a `zoomstart` event. `Map#flyTo` is a better example.

My apologies -- I know this is tangential to the point of this PR.

davidtheclark replied:
Perfect! I think of clarifying all these things as the point of the PR. It got pretty big, as you can see, so it's inevitable that I made typos and missed things like this. Thanks for looking things through in depth!

lucaswoj commented on js/ui/map.js:
```diff
@@ -57,48 +57,64 @@ var defaultOptions = {
+ *   bearing (rotation) will snap to north. For example, with a `bearingSnap` of 7, if the user has rotated
+ *   the map within 7 degrees of north, the map will automatically snap to north exactly.
+ * @param {Array<string>} [options.classes] Mapbox GL style class names with which to initialize the map.
+ *   Keep in mind that these are used for filtering a style layer's pain properties, so are *not* reflected
```
"for ~~filtering~~ a style layer's ~~pain~~ properties" -> "for **grouping** a style layer's **paint** properties"

lucaswoj commented on js/ui/map.js:
```diff
@@ -212,9 +228,13 @@ util.extend(Map.prototype, /** @lends Map.prototype */{
+     * Adds a Mapbox GL style class to the map.
      *
-     * @param {string} klass name of style class
+     * Keep in mind that these are used for filtering a style layer's pain properties, so are *not* reflected
```
paint properties

lucaswoj commented on js/ui/map.js:
```diff
@@ -267,30 +287,31 @@ util.extend(Map.prototype, /** @lends Map.prototype */{
+     * Resizes the map according to the dimensions of its
+     * `container` element.
+     *
+     * This method is often called after the map's `container` is resized by another script,
```
"is often" -> "must be"

lucaswoj commented on js/ui/map.js:
```diff
@@ -399,51 +428,56 @@ util.extend(Map.prototype, /** @lends Map.prototype */{
+     * @param {Object} [parameters]
+     * @param {Array<string>} [parameters.layers] An array of style layer IDs for the query to inspect.
+     *   Only features within these layers will be returned. If this parameter is undefined, all layers will be checked.
+     * @param {Array<string>} [parameters.filter] A [filter](https://www.mapbox.com/mapbox-gl-style-spec/#types-filter)
```
Let's leave this type parameter as `Array`. `Array<string>` doesn't accurately capture filters like `['any', ['==', '@penny', true], ['==', '@linus', true]]`

lucaswoj commented on js/ui/map.js:
```diff
@@ -548,11 +585,11 @@ util.extend(Map.prototype, /** @lends Map.prototype */{
-     * of a new style, not a URL string.
+     * Replaces the map's Mapbox GL style object with a new value.
+     *
+     * Unlike the `style` option in the `Map` constructor, this method only accepts an object, not a URL string.
```
Let's remove this line. `setStyle` does accept URL strings as the [normalization is handled within `Style`](https://github.com/mapbox/mapbox-gl-js/blob/master/js/style/style.js#L71-L75).

lucaswoj commented:
Done proofreading. I think we're ready to rebase & merge once the above feedback is taken into account. 

Thank you for tackling this project! Our docs are so much better off now.

davidtheclark commented:
I've made changes based on your feedback:

- Fixed typos you found.
- Altered indentation on the homepage’s example `<script>` tag.
- Added a little more explanation and an external link for GeoJSONSource’s `options.tolerance` and `options.buffer`. Also added defaults to these from what I saw in the code. *Are these explanations and defaults correct?* I am no expert yet ...
- Fixed the param types that you identified.
- Added `mouseout` event while rebasing.

I also found a few other little tweaks to perform as I read through some of it again.

I'm still uncertain about the event data, though. Specifically, those events that now have their object defined something like this: `@property {EventData} data The original event data, if the event was the result of user interaction.` Two questions:
- Can we just say this is `@property {EventData} data`, since the documentation for EventData covers the original event? 
- If on a simple example map I log the event data fired with `zoomend`, then I double-click to zoom, the object I see only has `type` (`"zoomend"`) and `target` (the `Map` instance) properties, not EventData's properties, and not original event data even though it originated from my double-click. What am I missing here?

lucaswoj commented:
> I'm still uncertain about the event data, though. Specifically, those events that now have their object defined something like this: `@property {EventData} data The original event data, if the event was the result of user interaction.` Two questions:
> - Can we just say this is `@property {EventData} data`, since the documentation for EventData covers the original event? 

Absolutely! I think this falls under the bullet point in the doc guidelines that says "Documentation descriptions must contain more information than what is obvious from the identifier and JSDoc metadata."

> - If on a simple example map I log the event data fired with `zoomend`, then I double-click to zoom, the object I see only has `type` (`"zoomend"`) and `target` (the `Map` instance) properties, not EventData's properties, and not original event data even though it originated from my double-click. What am I missing here?

This sounds like a bug to me.

lucaswoj commented:
While we're on the topic of `EventData`...

The names `Event` and `EventData` are not reflective of the types' semantics.

 - `Event` - any DOM event
 - `EventData` - any Mapbox GL JS event (some of which "wrap" DOM events)

Would renaming `EventData` to `MapboxEvent` make more sense?

davidtheclark commented:
> Would renaming EventData to MapboxEvent make more sense?

Yeah, that makes a lot of sense, I think.

So do *all* events documented here fire with a MapboxEvent? And lines like https://github.com/mapbox/mapbox-gl-js/pull/2755/files#diff-274e90d38623e2e16a5e7d34414f33c7R1325 are incorrect? (I tried this out, and it does look like with `map.on('mousedown', ..)` I get a MapboxEvent, not just the original DOM MouseEvent.)

lucaswoj commented:
It looks like some events have DOM `Event`s and some have `MapboxEvent`s. The intended distinction seems to be

 - events that proxy DOM Events (i.e. `mousedown`) have `Event`s
 - events that are unique to GL JS (i.e. `moveend`) have `MapboxEvent`s

lucaswoj commented:
Here's the code that fires `mousedown` https://github.com/mapbox/mapbox-gl-js/blob/master/js/ui/bind_handlers.js#L45-L49

bhousel commented:
> Would renaming EventData to MapboxEvent make more sense?

I don't think so.. `EventData` is just an unstructured bucket to store whatever the listener wants to watch for.   It's not an event.  It's not guaranteed to always contain the same stuff.  (e.g. the click events dispatch an eventData with a `LngLat` property, but boxzoom events dispatch an eventData with `LngLatBounds` property.)

davidtheclark commented:
@lucaswoj So I'm seeing in that code that the native `MouseEvent` from `mousedown` ends up being wrapped in an object matching the documented `EventData` API: https://github.com/mapbox/mapbox-gl-js/blob/master/js/ui/bind_handlers.js#L126. Seems, then, like we may never actually send the raw DOM `Event`?

> EventData is just an unstructured bucket to store whatever the listener wants to watch for

@bhousel Sounds, then, like the existing docs are inaccurate, right? They suggest that `EventData` is a class and we send out instances of it — when it fact it seems like it's more of an interface, with a few possible properties that aren't documented yet, like `LngLatBounds` for `boxzoom` or `LngLats` and `points` for touch events (https://github.com/mapbox/mapbox-gl-js/blob/master/js/ui/bind_handlers.js#L142)?

lucaswoj commented:
Sounds like we should replace all references to `EventData` with `Object`

bhousel commented:
> From what you're saying it sounds like the existing docs are inaccurate, right? They suggest that EventData is a class and we send out instances of it

Yes, I think that's right.  It's not so much a class, but more that the caller can pass an object with whatever tracing data they want to watch for, and our handlers will decorate that object with additional properties and dispatch on to any event listeners.  (see #1975, #1969 for context)

I don't know the best way to document it.  We could just call them all `Object` and then in each documentation block enumerate the extra properties that the decorated object might contain, but I thought that would get repetitive, so I rolled them all into a thing called `EventData`.

davidtheclark commented:
In d79732151b583dc2f7878f261e2c48415b1b44f7 I tried something to account for everything you've pointed out above. What do you think of that approach, @bhousel and @lucaswoj?

lucaswoj commented:
I'm learning so much about our event system! A huge thanks to @bhousel for all the help.

From what I understand, we can document our event system the most accurately with the following interfaces. Most event handlers could declare their signature as `Map*Event | Object` because they may return a `Map*Event` or an arbitrary user-provided object. (Declaring the signature as `Object` would be accurate too but providing a pointer to the relevant `Map*Event` is helpful to humans)

# `MapMouseEvent`

https://github.com/mapbox/mapbox-gl-js/blob/master/js/ui/bind_handlers.js#L126-L130 

```
/**
 * @typedef {Object} MapMouseEvent
 * @property {string} type The event type.
 * @property {Map} target The `Map` object that fired the event.
 * @property {MouseEvent} originalEvent
 * @property {Point} point The pixel coordinates of the mouse event target, relative to the map
 *   and measured from the top left corner.
 * @property {LngLat} lngLat The geographic location on the map of the mouse event target.
 */
```

# `MapTouchEvent`

https://github.com/mapbox/mapbox-gl-js/blob/master/js/ui/bind_handlers.js#L139-L145

```
/**
 * @typedef {Object} MapTouchEvent
 * @property {string} type The event type.
 * @property {Map} target The `Map` object that fired the event.
 * @property {TouchEvent} originalEvent
 * @property {Point} point The pixel coordinates of the center of the touch event points, relative to the map
 *   and measured from the top left corner.
 * @property {LngLat} lngLat The geographic location on the map of the center of the touch event points.
 */
```

# `MapBoxZoomEvent`

```
/**
 * @typedef {Object} MapTouchEvent
 * @property {LngLatBounds} boxZoomBounds The bounding box of the "box zoom" interaction.
 * @property {MapMouseEvent} originalEvent
 */
```

davidtheclark commented:
Great! I will update the PR soon.

> On Jun 30, 2016, at 12:32 PM, Lucas Wojciechowski <notifications@github.com> wrote:
> 
> I'm learning so much about our event system! A huge thanks to @bhousel for all the help.
> 
> From what I understand, we can document our event system the most accurately with the following interfaces. Most event handlers could declare their signature as Map*Event | Object because they may return a Map*Event or an arbitrary user-provided object. (Declaring the signature as Object would be accurate too but providing a pointer to the relevant Map*Event is helpful to humans)
> 
> MapMouseEvent
> 
> https://github.com/mapbox/mapbox-gl-js/blob/master/js/ui/bind_handlers.js#L126-L130
> 
> /**
>  * @typedef {Object} MapMouseEvent
>  * @property {string} type The event type.
>  * @property {Map} target The `Map` object that fired the event.
>  * @property {MouseEvent} originalEvent
>  * @property {Point} point The pixel coordinates of the mouse event target, relative to the map
>  *   and measured from the top left corner.
>  * @property {LngLat} lngLat The geographic location on the map of the mouse event target.
>  */
> MapTouchEvent
> 
> https://github.com/mapbox/mapbox-gl-js/blob/master/js/ui/bind_handlers.js#L139-L145
> 
> /**
>  * @typedef {Object} MapTouchEvent
>  * @property {string} type The event type.
>  * @property {Map} target The `Map` object that fired the event.
>  * @property {TouchEvent} originalEvent
>  * @property {Point} point The pixel coordinates of the center of the touch event points, relative to the map
>  *   and measured from the top left corner.
>  * @property {LngLat} lngLat The geographic location on the map of the center of the touch event points.
>  */
> MapBoxZoomEvent
> 
> /**
>  * @typedef {Object} MapTouchEvent
>  * @property {LngLatBounds} boxZoomBounds The bounding box of the "box zoom" interaction.
>  * @property {MapMouseEvent} originalEvent
>  */
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

davidtheclark commented:
Looks like `zoom` and `move` events typically are just `type`, `target`, and `originalEvent` (either `MouseEvent` or `TouchEvent`), unless they are caused by a double-click zoom, in which case there's no `originalEvent`.

lucaswoj commented:
I am so excited! :shipit:

lucaswoj commented:
Do you want to merge this into `mb-pages` instead of `master`?

davidtheclark commented:
Sure, I can rebase onto that and submit a new PR.

What do you think of the ambiguity about `zoom` and `move` events?

lucaswoj commented:
> Sure, I can rebase onto that and submit a new PR.

👍 Feel free to merge that one as soon as CI passes.

> What do you think of the ambiguity about zoom and move events?

I think the way you've handled it is good. We can make the implementation more consistent in a follow-up task.

lucaswoj commented:
I just made a PR to fix the inconsistency in the `DoubleClickZoomHandler`

davidtheclark commented:
Moved this to #2815.
